src/libsync/platform_mac.mm

```diff
@@ -39,20 +44,115 @@ - (BOOL)applicationShouldHandleReopen:(NSApplication *)sender hasVisibleWindows:
+    }
+
+private:
+    static void sleepWakeupCallBack(void *refParam, io_service_t service, natural_t messageType, void *messageArgument)
+    {
+        Q_UNUSED(service)
+
+        auto thiz = static_cast<PowerNotificationsListener *>(refParam);
+
+        switch (messageType) {
+        case kIOMessageCanSystemSleep:
+            /* Idle sleep is about to kick in. This message will not be sent for forced sleep.
+             * Applications have a chance to prevent sleep by calling IOCancelPowerChange.
+             * Most applications should not prevent idle sleep. Power Management waits up to
+             * 30 seconds for you to either allow or deny idle sleep. If you don’t acknowledge
+             * this power change by calling either IOAllowPowerChange or IOCancelPowerChange,
+             * the system will wait 30 seconds then go to sleep.
+             */
+
+            qCInfo(OCC::lcPlatform) << "System power message: can system sleep?";
+
+            // Uncomment to cancel idle sleep
+            // IOCancelPowerChange(thiz->rootPowerDomain, reinterpret_cast<long>(messageArgument));
+
+            // Allow idle sleep
+            IOAllowPowerChange(thiz->rootPowerDomain, reinterpret_cast<long>(messageArgument));
+            break;
+
+        case kIOMessageSystemWillNotSleep:
+            /* Announces that the system has retracted a previous attempt to sleep; it
+             * follows `kIOMessageCanSystemSleep`.
+             */
+            qCInfo(OCC::lcPlatform) << "System power message: system will NOT sleep.";
+            break;
+
+        case kIOMessageSystemWillSleep:
+            printf("kIOMessageSystemWillSleep message.\n");
+            /* The system WILL go to sleep. If you do not call IOAllowPowerChange or
+             * IOCancelPowerChange to acknowledge this message, sleep will be delayed by
+             * 30 seconds.
+             *
+             * NOTE: If you call IOCancelPowerChange to deny sleep it returns kIOReturnSuccess,
+             * however the system WILL still go to sleep.
+             */
+
+            qCInfo(OCC::lcPlatform) << "System power message: system WILL sleep.";
+
+            IOAllowPowerChange(thiz->rootPowerDomain, reinterpret_cast<long>(messageArgument));
+            break;
+
+        case kIOMessageSystemWillPowerOn:
+            /* Announces that the system is beginning to power the device tree; most devices
+             * are still unavailable at this point.
+             */
+            qCInfo(OCC::lcPlatform) << "System power message: system will power on.";
+            break;
+
+        case kIOMessageSystemHasPoweredOn:
+            /* Announces that the system and its devices have woken up. */
+            qCInfo(OCC::lcPlatform) << "System power message: system has powered on.";
+            break;
+
+        default:
+            /* Not a system sleep and wake notification. */
```

Comment:
I guess we should handle all cases and not define default to make clang tell us about unhanded cases.
Additionally we should not use printf

Reply:
There are more messages, some even unused. See https://opensource.apple.com/source/xnu/xnu-2050.48.11/iokit/IOKit/IOMessage.h.auto.html . I think a break here is in order, also because Apple might add new ones. Also, take a look if we want to receive/log other events.

Reply:
(printfs have been removed)

Reply:
Ok I expected it to be a enum, with defines no compiler in the world will be able to warn for new values.
Can we then log the unhandled cases to debug?